rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,1053 @@
+
+These APIs have been prototyped internally and refined in collaboration
+with Alphabet researchers. We hope that they will be useful for many TensorFlow users. This use-case will be the main focus of this design document.
+
+#### Use case #3: Users who will create custom distributed layers
+
+Most of the users that write their custom outer training loops will rely on common TensorFlow components to handle communication between replicated computation. For instance, TensorFlow optimizers will automatically take care of aggregating the gradients across replicas. Some users, however, may have complex models where they want to do custom distributed communication in the middle of their model, e.g. insert an all-reduce operation.
+
+For example, some users may want to write a custom layer to do cross-replica batch normalization. For such users, we provide additional methods such as all-reduce and broadcast to facilitate communication between replicas. Since these use cases are uncommon, we will abstract away these methods under an `extended` property of the main DistributionStrategy class.
+
+#### Use case #4: Users who will create new `DistributionStrategy` implementations
+
+We hope that some users will extend the `DistributionStrategy` class and create new implementations. For instance, authors of [Horovod](https://github.com/uber/horovod) are collaborating with TensorFlow team to create a new `HorovodDistributionStrategy`. These users of course will need to understand and implement the entire surface area of the Distribution Strategy API. 
+
+
+
+## Integration with tf.keras and Estimator
+
+As mentioned before, the majority of TensorFlow users will use Distribution Strategy through one of the high level training frameworks - `tf.keras` and `Estimator`. Users can distribute their code written with these APIs to GPUs / TPUs / multiple machines with typically only 2-3 lines of code changes. We’ve modified the `tf.keras` and `Estimator` backends, as well as many TensorFlow components (such as optimizers, metrics, and layers), to be Distribution Strategy aware. The user only needs to decide which strategy type they want to use for their platform. The rest is taken care of by the framework.
+
+This usage has been covered in the Distribution Strategy [README document] (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/distribute/README.md). Here is a simple example of scaling training in tf.keras to multiple GPUs:
+
+```python
+inputs = tf.keras.layers.Input(shape=(1,))
+predictions = tf.keras.layers.Dense(1)(inputs)
+model = tf.keras.models.Model(inputs=inputs, outputs=predictions)
+dataset = tf.data.Dataset.zip((features, labels))
+strategy = tf.distribute.MirroredStrategy()
+model.compile(loss='mean_squared_error',
+              optimizer=tf.train.GradientDescentOptimizer(learning_rate=0.2),
+              distribute=strategy)
```

Comment:
The plan for input is to rely on datasets to be able to (eventually) efficiently shard input. And this can happen at the compile/fit stage, with whatever dataset is provided at that time. Note that the initial implementation may not be efficient, but we want to design the APIs such that it makes the most sense for users, even if we have to sacrifice performance in the short term. 

AFAIK, there aren't any other limitations we've seen so far because of delaying strategy passing until compile that affect performance, but @anj-s maybe you can add more thoughts on this?

Reply:
I believe compile is the right stage to pass the strategy and set it as one of the model properties, similar to loss, optimizer etc.  We also only have access to the input at the fit stage (and possibly information about the workers as well). There isn't any performance issues so far with the current flow. As guptapriya@ mentioned we will be using an initial implementation for input sharding which may have performance issues but this is not dependent on when we pass the strategy object.